ListMap update order is the same as VectorMap (scala/bug#11306)

joshlemer opened the pull request:
Fixes scala/bug#11306

But depends upon https://github.com/scala/scala/pull/7509

joshlemer committed:
[11306] ListMap update order conforms to that of VectorMap

szeiger commented:
Simpler alternative in https://github.com/scala/scala/pull/7581 so this needs some benchmarking to see if the added complexity is worth it.

joshlemer commented:
@szeiger 
Benchmarking code:
```scala
package scala.collection.immutable

import org.openjdk.jmh.annotations._
import org.openjdk.jmh.infra._
import org.openjdk.jmh.runner.IterationType
import benchmark._
import java.util.concurrent.TimeUnit

@BenchmarkMode(Array(Mode.AverageTime))
@Fork(2)
@Threads(1)
@Warmup(iterations = 10)
@Measurement(iterations = 10)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@State(Scope.Benchmark)
class ListMapBenchmark {
  @Param(Array("1", "10", "100", "1000", "10000", "100000"))//, "1000000"))
  var size: Int = _

  var listMap: ListMap[Int, Int] = _

  @Setup(Level.Trial)
  def initKeys(): Unit = {
    listMap = (0 until size).map(_ -> 0).to(ListMap)
  }
  @Benchmark
  def updateWithReplacementInMiddleOfListMap(bh: Blackhole): Unit = {
    // must traverse to the middle of the listmap,
    // discover that the value must be updated, rebuild a new ListMap with the new value
    bh.consume(listMap.updated(size/2, 1))
  }
  @Benchmark
  def updateWithNoReplacementNeededInMiddleOfListMap(bh: Blackhole): Unit = {
    // must traverse to the middle of the listmap,
    // discover that no update is needed (value is not changed)
    bh.consume(listMap.updated(size/2, 0))
  }
}
```

This solution's results:

```
[info] # Run complete. Total time: 00:09:11
[info] Benchmark                                                        (size)  Mode  Cnt       Score       Error  Units
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap       1  avgt   20       3.909 ±     0.083  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap      10  avgt   20       9.978 ±     0.471  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap     100  avgt   20      95.064 ±     6.123  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap    1000  avgt   20    1000.978 ±    13.473  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap   10000  avgt   20   11346.854 ±   145.063  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap  100000  avgt   20  116322.796 ±  3411.073  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap               1  avgt   20       6.416 ±     0.044  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap              10  avgt   20      28.779 ±     0.602  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap             100  avgt   20     267.658 ±     3.344  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap            1000  avgt   20    2877.489 ±    48.095  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap           10000  avgt   20   36857.151 ±   631.221  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap          100000  avgt   20  351977.263 ± 11207.433  ns/op
```

Other solution results

```
[info] Benchmark                                                        (size)  Mode  Cnt       Score        Error  Units
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap       1  avgt   20       8.026 ±      1.225  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap      10  avgt   20      54.547 ±      2.965  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap     100  avgt   20     617.124 ±     43.185  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap    1000  avgt   20    4192.013 ±     37.891  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap   10000  avgt   20   45940.102 ±    554.233  ns/op
[info] ListMapBenchmark.updateWithNoReplacementNeededInMiddleOfListMap  100000  avgt   20  445130.330 ±  36781.760  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap               1  avgt   20       6.653 ±      0.656  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap              10  avgt   20      56.190 ±      1.918  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap             100  avgt   20     587.042 ±     10.654  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap            1000  avgt   20    5102.368 ±    777.422  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap           10000  avgt   20   49369.585 ±   2806.469  ns/op
[info] ListMapBenchmark.updateWithReplacementInMiddleOfListMap          100000  avgt   20  666655.868 ± 142530.812  ns/op

```

szeiger reviewed (approved):
The performance numbers look convincing.
